src/v/cluster/controller_backend.cc

```diff
@@ -582,24 +591,68 @@ controller_backend::calculate_learner_initial_offset(
     /**
      * Calculate retention targets based on cluster and topic configuration
      */
-    const auto initial_retention_bytes = get_topic_property(
+    auto initial_retention_bytes = get_topic_property(
       _initial_retention_local_target_bytes(),
       log->config().has_overrides()
         ? log->config().get_overrides().initial_retention_local_target_bytes
         : tristate<size_t>{std::nullopt});
 
-    const auto initial_retention_ms = get_topic_property(
+    auto initial_retention_ms = get_topic_property(
       _initial_retention_local_target_ms(),
       log->config().has_overrides()
         ? log->config().get_overrides().initial_retention_local_target_ms
         : tristate<std::chrono::milliseconds>{std::nullopt});
+
     /**
-     * Initial target retention disabled
+     * There are two possibilities for learner start offset calculation:
+     *
+     * >>> fast partition movement <<<
+     * - the reconfiguration policy is set to use target_initial_retention and
+     *   initial retention is configured, in this case the initial learner
+     *   offset will be calculated based on the initial target retention
+     *   settings
+     *
+     * >>> full local retention move <<<
+     * - with non strict local retention the storage manager may allow
+     *   partitions to grow beyond their configured local retention target. In
+     *   this case the controller backend will use the local retention target
+     *   properties and will schedule move delivering only the data that would
+     *   be retained if local retention was working in strict mode regardless of
+     *   initial retention settings and configured move policy.
      */
-    if (
-      !initial_retention_bytes.has_value()
-      && !initial_retention_ms.has_value()) {
-        return std::nullopt;
+    const bool no_initial_retention_settings = !(
```

Comment:
nit: feels having a bit off having all this code in controller_backend. I wonder if we should encapsulate it inside `disk_log_impl` where every relevant bit of info (including default retention settings) is readily available.

Reply:
maybe i can refactor it in a follow, this will be easier to backport

Reply:
> nit: feels having a bit off having all this code in controller_backend. I wonder if we should encapsulate it inside disk_log_impl where every relevant bit of info (including default retention settings) is readily available.

good point. it's already hard enough parsing this sort of logic in disk_log_impl now it's here too :(